### PR TITLE
linker: Introduce zephyr,memory-region compatible

### PR DESCRIPTION
**Pull request by carlocaione**
Introduce a new `zephyr,memory-region` compatible to be used when a new memory region must be created in the linker script from the devicetree node using the compatible.

Remove also the `LINKER_DT_REGION_FROM_NODE` macro and add a new `LINKER_DT_REGION` macro to cycle through all the compatible regions.

This is a preparatory work to be integrated with #41223 so that under one single compatible we have all the properties to create a new region in the linker script and set the MPU properties of the region using the info coming from the DT.

### CODE REVIEW

**Comment by JordanYates** [comment]
AFAICT we've lost the `rw` or `rwx` attributes on the memory regions with this PR.
I suspect we'll need an additional property called `zephyr,memory-region-attr` or similar

**Comment by carlocaione** [comment]
> AFAICT we've lost the `rw` or `rwx` attributes on the memory regions with this PR.
> I suspect we'll need an additional property called `zephyr,memory-region-attr` or similar

Yes, but do we _really_ need those? AFAICT those are attributes are optional and they will conflict with the MPU configuration parameter that I want to add later, so I would gladly leave that out if possible (how are used by the toolchain?)

**Comment by JordanYates** [comment]
> Yes, but do we really need those? AFAICT those are attributes are optional and they will conflict with the MPU configuration parameter that I want to add later, so I would gladly leave that out if possible (how are used by the toolchain?)

Based on the description here, https://sourceware.org/binutils/docs/ld/MEMORY.html, it looks like they're only used as a fallback of where to place sections that are not explicitly placed. All of our sections should be explicitly placed AFAIK and the `FLASH` and `SRAM` regions still have their attributes, so I'm guessing it should be fine.

**Review by JordanYates — DISMISSED** [review]

File: include/linker/devicetree_regions.h
Comment (mbolivar-nordic) [inline include/linker/devicetree_regions.h:59]: I think it would be clearer to call this `LINKER_DT_REGIONS`, since it declares multiple regions, not just one region.

**Review by mbolivar-nordic — COMMENTED** [review]
No objections from the devicetree side if this works for both you and @JordanYates . I'd like @tejlmand to take a look at the build system portion since I don't see support on the armclang side.

One minor naming nit.

**Comment by mbolivar-nordic** [comment]
> No objections from the devicetree side if this works for both you and @JordanYates

I also agree with your comment here (https://github.com/zephyrproject-rtos/zephyr/pull/41223#issuecomment-1018270948) that @Nukersson should take a look at this before merging too; sorry for missing that at first.

**Comment by Nukersson** [comment]
> > AFAICT we've lost the `rw` or `rwx` attributes on the memory regions with this PR.
> > I suspect we'll need an additional property called `zephyr,memory-region-attr` or similar
> 
> Yes, but do we _really_ need those? AFAICT those are attributes are optional and they will conflict with the MPU configuration parameter that I want to add later, so I would gladly leave that out if possible (how are used by the toolchain?)

@JordanYates @carlocaione I think, leaving attributes away breaks the interface. Can the attributes stay untouched? Is it technically difficult to implement?

File: soc/riscv/riscv-privilege/andes_v5/ae350/linker.ld
Comment (Nukersson) [inline soc/riscv/riscv-privilege/andes_v5/ae350/linker.ld:90]:
```suggestion
    LINKER_DT_REGION(rw)
```

File: include/arch/arm/aarch32/cortex_a_r/scripts/linker.ld
Comment (Nukersson) [inline include/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:84]:
```suggestion
    LINKER_DT_REGION(rw)
```

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld
Comment (Nukersson) [inline include/arch/arm/aarch32/cortex_m/scripts/linker.ld:84]:
```suggestion
    LINKER_DT_REGION(rw)
```
1. Attributes are not used. 2. Just ignore rwx and use rw for all since not used in later implementation.

File: include/arch/riscv/common/linker.ld
Comment (Nukersson) [inline include/arch/riscv/common/linker.ld:84]:
```suggestion
    LINKER_DT_REGION(rw)
```

File: include/linker/devicetree_regions.h
Comment (Nukersson) [inline include/linker/devicetree_regions.h:49]:
```suggestion
#define _REGION_DECLARE(node_id, attr)	      \
	LINKER_DT_NODE_REGION_NAME(node_id)(attr) : \
```

File: include/linker/devicetree_regions.h
Comment (Nukersson) [inline include/linker/devicetree_regions.h:59]:
```suggestion
#define LINKER_DT_REGION(attr) \
```

File: include/linker/devicetree_regions.h
Comment (Nukersson) [inline include/linker/devicetree_regions.h:60]:
```suggestion
	DT_FOREACH_STATUS_OKAY(_DT_COMPATIBLE, _REGION_DECLARE(attr))
```
I am not sure here, if it is correct suggestion, how to use attributes in this context.

**Review by KozhinovAlexander — CHANGES_REQUESTED** [review]
@carlocaione Great work and respect for pushing this very important topic forward!

Please add  take a look on my review and changes and agree or disagree with proof. I hope, it is not a huge amount work to do for you 😃

**Comment by JordanYates** [comment]
> The 2nd should accept an array (ideally) or proposed zephyr,memory-region-attr. But you're lucky here, because the attributes are same for all the regions and not used 😄. Thus yo can use one attribute only for sacke of simplicity. Later somebody else can add the array, if there would be a need.

That is not the case currently, the `ti-ccfg` region uses `rwx`

**Comment by Nukersson** [comment]
I've proposed a change with comment, why rw can be done instead of rwx. Please take a look on it.

**Comment by Nukersson** [comment]
@carlocaione Well, you do not have to implement attributes propagation to MPU. Just leave them where they were and put comment on the macro, that attributes are not used and will stay here for compatibility reasons. I think, there should be nothing thrown away.

**Comment by carlocaione** [comment]
> @carlocaione Well, you do not have to implement attributes propagation to MPU. Especially you do not have to use them in your code. Just leave them as arguments of the macros as they are.

Ok, if I do not have to propagate the attributes to the MPU, I do not have to use them in the code, they are not used anywhere, why are we bothering to leave those in place?

> Please put a comment on the macro, that attributes are not used and will stay here for compatibility reasons. I think, there should be nothing thrown away.

Compatibility with what exactly?

**Comment by Nukersson** [comment]
Obviously attributes are a dead code, I agree with you here. But we do not know for what the were originally introduced, since we do not have documentation. Therefore it is better to leave them at place, since there is no proof, they were introduced for some reason. Just do not delete old code or maybe not functional code, if you do not exactly know, where it can hit.

**Comment by carlocaione** [comment]
> Obviously attributes are a dead code, I agree with you here. But we do not know for what they were originally introduced, since we do not have documentation.

The documentation is clear on what the attributes are for (see https://sourceware.org/binutils/docs/ld/MEMORY.html) and for what Zephyr is concerned we do not have an actual use for those.

> Therefore it is better to leave them at place, since there is no proof, they were introduced for some reason. Just do not delete old code or maybe not functional code, if you do not exactly know, where it can hit.

This is not possible. We have two ways here:
- We remove the attributes because there is no use for those and they will conflict with the MPU configuration
- We fully support the attributes because I do not want to introduce technical debt. That means that we have to support all the possible combinations (`r`, `rw`, `rwx`, etc...) by adding a new DT property and we have to find a way to make this co-exist with MPU configuration.

> Effectively, from my side I would give this PR my approve also with deleted attributes, cause it is a good working PR.

I'll add a comment explaining why the attributes are missing, thanks.

**Review by katsuster — DISMISSED** [review]

File: include/linker/devicetree_regions.h
Comment (gmarull) [inline include/linker/devicetree_regions.h:10]: nit: internal helper macros should all be put under

```
/** @cond INTERNAL_HIDDEN */
...
/** @endcond */
```

**Review by gmarull — DISMISSED** [review]
One minor documentation comment

File: include/linker/devicetree_regions.h
Comment (carlocaione) [reply include/linker/devicetree_regions.h:10]: In that file there is already an `INTERNAL_HIDDEN` section, so I'm just moving the `_DT_COMPATIBLE` in there.

**Comment by carlocaione** [comment]
@tejlmand can you take a look?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Great work.

We need to be able to be able to fetch all `zephyr,memory-region` compatibles in same way here:
https://github.com/zephyrproject-rtos/zephyr/blob/a03deb8ef3c940bac125e35442ba00bd2e6f51bb/cmake/linker_script/arm/linker.cmake#L37-L50

so that code can use same principle.

Everything else looks good.
Let me know if you would like me to take a look at it.

**Comment by carlocaione** [comment]
@tejlmand I gave it a shot.

@mbolivar-nordic you want probably take a look at this again.

File: cmake/extensions.cmake
Comment (tejlmand) [inline cmake/extensions.cmake:2848]: Function is called: `dt_comp_path()`

```suggestion
      message(FATAL_ERROR "dt_comp_path(${ARGV0} ...) "
```

File: soc/riscv/riscv-privilege/andes_v5/ae350/linker.ld
Comment (tejlmand) [inline soc/riscv/riscv-privilege/andes_v5/ae350/linker.ld:90]: When testing with `arty_a7_arm_designstart_m3` this code results in `DTCM` and `ITCM` being defined on the same line like this:

```
MEMORY
    {
    FLASH (rx) : ORIGIN = (0x0 + 0x0), LENGTH = (32*1K - 0x0)
    SRAM (wx) : ORIGIN = 0x20000000, LENGTH = (32 * 1K)
    "ITCM" : ORIGIN = 0, LENGTH = 32768 "DTCM" : ORIGIN = 536870912, LENGTH = 32768
    IDT_LIST (wx) : ORIGIN = 0xFFFFF7FF, LENGTH = 2K
    }
```

it's not invalid code, but would be nice if each could have their own line.

This looks very nice when using `-DCONFIG_CMAKE_LINKER_GENERATOR=y` :wink:

**Review by tejlmand — COMMENTED** [review]
Looks very good, thanks for putting this together.

A couple of very minor nits.

File: cmake/extensions.cmake
Comment (carlocaione) [reply cmake/extensions.cmake:2843]: yup

File: cmake/extensions.cmake
Comment (carlocaione) [reply cmake/extensions.cmake:2848]: right

File: cmake/linker_script/arm/linker.cmake
Comment (carlocaione) [reply cmake/linker_script/arm/linker.cmake:41]: ;)

File: scripts/dts/gen_dts_cmake.py
Comment (mbolivar-nordic) [inline scripts/dts/gen_dts_cmake.py:99]: Naming nitpick here. `iter` is a built-in python function that returns an iterator over an iterable. So the name `comp_iter` suggests it is an iterator. But it's not -- it's a dict, which is not an iterator.

I suggest renaming this `compatible2paths` to match the general edtlib convention of naming a dict `<key>2<value>`

**Review by mbolivar-nordic — APPROVED** [review]
I'm happy with the devicetree portions; @tejlmand should decide about the build system parts.

One nonblocking suggestion for the script.

Thanks for your persistence, @carlocaione.

File: scripts/dts/gen_dts_cmake.py
Comment (carlocaione) [reply scripts/dts/gen_dts_cmake.py]: yup, done.

File: soc/riscv/riscv-privilege/andes_v5/ae350/linker.ld
Comment (tejlmand) [reply soc/riscv/riscv-privilege/andes_v5/ae350/linker.ld:90]: I agree, not easy at all and I can't think of a solution, just wanted to mention it.

**Review by tejlmand — APPROVED** [review]
Looks good.

Thanks for the effort.

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**
